ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/NettyPerfClient.java

```diff
@@ -0,0 +1,501 @@
+      long currentChunkReceiveTime = System.currentTimeMillis();
+      boolean recognized = false;
+      if (in instanceof HttpResponse) {
+        recognized = true;
+        long responseReceiveStart = currentChunkReceiveTime - requestStartTime;
+        perfClientMetrics.timeToFirstResponseChunkInMs.update(responseReceiveStart);
+        logger.trace("Response receive has started on channel {}. Took {} ms", ctx.channel(), responseReceiveStart);
+        response = (HttpResponse) in;
+      }
+      if (in instanceof HttpContent) {
+        recognized = true;
+        perfClientMetrics.delayBetweenChunkReceiveInMs.update(currentChunkReceiveTime - lastChunkReceiveTime);
+        chunksReceived++;
+        int bytesReceivedThisTime = ((HttpContent) in).content().readableBytes();
+        sizeReceived += bytesReceivedThisTime;
+        perfClientMetrics.bytesReceiveRate.mark(bytesReceivedThisTime);
+        if (in instanceof LastHttpContent) {
+          long requestRoundTripTime = currentChunkReceiveTime - requestStartTime;
+          perfClientMetrics.requestRoundTripTimeInMs.update(requestRoundTripTime);
+          perfClientMetrics.getContentSizeInBytes.update(sizeReceived);
+          perfClientMetrics.getChunkCount.update(chunksReceived);
+          logger.trace(
+              "Final content received on channel {}. Took {} ms. Total chunks received - {}. Total size received - {}",
+              ctx.channel(), requestRoundTripTime, chunksReceived, sizeReceived);
+          if (HttpHeaders.isKeepAlive(response) && isRunning) {
+            logger.trace("Sending new request on channel {}", ctx.channel());
+            sendRequest(ctx);
+          } else if (!isRunning) {
+            logger.info("Closing channel {} because NettyPerfClient has been shutdown", ctx.channel());
+            ctx.close();
+          } else {
+            perfClientMetrics.requestResponseError.inc();
+            logger.error("Channel {} not kept alive. Last response status was {}", ctx.channel(), response.getStatus());
+            ctx.close();
+          }
+        }
+      }
+      if (!recognized) {
+        throw new IllegalStateException("Unexpected HttpObject - " + in.getClass());
+      }
+      lastChunkReceiveTime = currentChunkReceiveTime;
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) {
+      logger.trace("Channel {} inactive", ctx.channel());
+      ctx.close();
+      if (isRunning) {
+        perfClientMetrics.unexpectedDisconnectionError.inc();
+        logger.info("Creating a new channel to keep up concurrency");
+        b.connect().addListener(channelConnectListener);
+      }
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+      perfClientMetrics.requestResponseError.inc();
+      logger.error("Exception caught on channel {} while processing request/response", ctx.channel(), cause);
+      ctx.close();
```

Comment:
After the `ctx.close()`, `channelInactive()` will be called eventually. `channelInactive()` is called irrespective of either side (server or client) closing the connection.